plugin-server/src/logs-ingestion/services/logs-rate-limiter.service.ts

```diff
@@ -0,0 +1,135 @@
+}
+
+export type FilteredMessages = {
+    allowed: LogsIngestionMessage[]
+    dropped: LogsIngestionMessage[]
+}
+
+/**
+ * The LogsRateLimiterService is used to rate limit logs ingestion to ensure we aren't allowing too many logs to be ingested at once.
+ * The key part is we specify the refill rate as our per second KB/s limit. and the bucket size as the amount we are allowed to burst to.
+ * The burst shouldn't be too much higher.
+ */
+export class LogsRateLimiterService {
+    constructor(
+        private hub: Hub,
+        private redis: RedisV2
+    ) {}
+
+    private rateLimitArgs(id: string, cost: number): [string, number, number, number, number, number] {
+        const nowSeconds = Math.round(Date.now() / 1000)
+
+        return [
+            `${REDIS_KEY_TOKENS}/${id}`,
+            nowSeconds,
+            cost,
+            this.hub.LOGS_LIMITER_BUCKET_SIZE_KB,
+            this.hub.LOGS_LIMITER_REFILL_RATE_KB_PER_SECOND,
+            this.hub.LOGS_LIMITER_TTL_SECONDS,
+        ]
+    }
+
+    public async rateLimitMany(idCosts: [string, number][]): Promise<[string, LogsRateLimit][]> {
+        const res = await this.redis.usePipeline({ name: 'logs-rate-limiter', failOpen: true }, (pipeline) => {
+            idCosts.forEach(([id, cost]) => {
+                pipeline.checkRateLimitV2(...this.rateLimitArgs(id, cost))
+            })
+        })
+
+        if (!res) {
+            throw new Error('Failed to rate limit')
+        }
+
+        return idCosts.map(([id], index) => {
+            const [tokenRes] = getRedisPipelineResults(res, index, 1)
+            const tokensBefore = Number(tokenRes[1]?.[0] ?? this.hub.LOGS_LIMITER_BUCKET_SIZE_KB)
+            const tokensAfter = Number(tokenRes[1]?.[1] ?? this.hub.LOGS_LIMITER_BUCKET_SIZE_KB)
+            return [
+                id,
+                {
+                    tokensBefore,
+                    tokensAfter,
+                    isRateLimited: tokensAfter <= 0,
+                },
+            ]
+        })
+    }
+
+    private isRateLimitingEnabledForTeam(teamId: number): boolean {
+        const enabledTeams = this.hub.LOGS_LIMITER_ENABLED_TEAMS
+        if (enabledTeams === '*') {
+            return true
+        }
+        if (!enabledTeams) {
+            return false
+        }
+        const teamIds = enabledTeams.split(',').map((id) => parseInt(id.trim(), 10))
+        return teamIds.includes(teamId)
+    }
+
+    public async filterMessages(messages: LogsIngestionMessage[]): Promise<FilteredMessages> {
+        // Group messages by team to calculate total cost per team (only for teams with rate limiting enabled)
+        const teamCosts = new Map<number, number>()
+        for (const message of messages) {
+            if (!this.isRateLimitingEnabledForTeam(message.teamId)) {
+                continue
+            }
+            const currentCost = teamCosts.get(message.teamId) ?? 0
+            // Cost is in KB (uncompressed bytes / 1024)
```

Comment:
i wonder if we should use KB/MB instead of KiB/MiB

I feel like more and more the SI units are standard and base 2 is dying, might be more obvious for customers?

Then again, base2 is more generous and we might get complaints about throttling at 0.95MiB